components/vision.py

```diff
@@ -99,42 +99,41 @@ def execute(self) -> None:
+
         # Get vision pose estimate
         # work out where the vision data was taken from based on histories
-        camera_pose = self.get_vis_pose_at(self.vision_data.timestamp)
-        # angle from target to robot in world space
-        angle_from_target = constrain_angle(
-            camera_pose.rotation().radians() - self.vision_data.angle + math.pi
-        )
-        # work out where vision though it was when the image was taken
-        vis_estimate = Translation2d(
-            distance=self.vision_data.distance,
-            angle=Rotation2d(angle_from_target),
-        )
-        self.field_obj.setPose(
-            goal_to_field(Pose2d(vis_estimate, self.imu.getRotation2d()))
+        t = self.vision_data.timestamp
+        turret_angle = self.turret.get_angle_at(t)
+        chassis_heading = self.chassis.get_pose_at(t).rotation().radians()
+
+        # Ranges from vision are from centre of goal to the camera
+        # Add the offset to get range to the centre of the turret
+        range = self.vision_data.distance + self.CAMERA_OFFSET
+
+        vision_pose = pose_from_vision(
+            range, turret_angle + self.vision_data.angle, chassis_heading
         )
-        # calcualte vision std dev
-        # trust vision less the more outdated it is
-        # vis_age = wpilib.Timer.getFPGATimestamp() - self.vision_data.timestamp
-        # age_fit = max(0, scale_value(vis_age, 0, 0.2, 1, 0))
-        # # trust vision less the more it thinks we've moved (to reduce impact of false positives)
-        # innovation = vis_estimate.distance(
-        #     self.chassis.estimator.getEstimatedPosition().translation()
-        # )
-        # # will be 0 if innovation is over 1.5m
-        # innovation_fit = min(1, max(0, scale_value(innovation, 0.25, 1.5, 1, 0)))
-        # # combined vision confidence is 0-1
-        # vis_confidence = self.vision_data.fittness * innovation_fit * age_fit
-        # if vis_confidence > 0.4:
-        #     vis_std_dev = 0.5 / vis_confidence
-        #     # pass vision pose estimate to chassis kalman filter
-        #     self.chassis.estimator.addVisionMeasurement(
-        #         Pose2d(vis_estimate, self.imu.getRotation2d()),
-        #         self.vision_data.timestamp,
-        #         [vis_std_dev, vis_std_dev, 0.0],
-        #     )
-        self.last_data_timestamp = self.vision_data.timestamp
+
+        self.field_obj.setPose(goal_to_field(vision_pose))
+
+        if self.fuse_vision_observations:
+            innovation = vision_pose.translation().distance(
+                self.chassis.estimator.getEstimatedPosition().translation()
+            )
+            # Gate on innovation
+            if innovation > 1.0:
+                return
+            # Come up with a position std dev from the angle in the picture
+            # When the target is near the edge, the estimate of range is worse
+            pos_std_dev = 0.2 + 0.4 * abs(self.vision_data.angle) / math.radians(15)
```

Comment:
do we want to incorporate the confidence returned from vision in this std dev?

Reply:
I thought about that, but it isn't actually in a scale that is useful. There is no way from looking at the vision code to work out what the range of values it returns are, so I ignored it and did it this way.
Ideally the vision system should return a confidence for the values it returns, and these should be in the same units as the observations. @LucienMorey @whert-dev ?

Reply:
Okay, so now it uses the fitness (which is in the range [0,1] to derive a std dev.